libs/data/bbhnet/data/dataloader.py

```diff
@@ -0,0 +1,372 @@
+                a `StopIteration` while iterating
+            waveform_sampler:
+                An object with `.fit` and `.sample` attributes for
+                sampling interferometer responses to raw waveforms
+                for injection into background data. If left as `None`,
+                no injection will take place at data loading time and
+                `waveform_frac` must be set to 0.
+            waveform_frac:
+                The fraction of each batch that should consist
+                of injected waveforms, and be marked with a
+                `1.` in the target tensor produced during iteration.
+            min_snr:
+                Minimum SNR value for sampled waveforms. Must be
+                specified if `waveform_dataset` is not `None`.
+            max_snr:
+                Maximum SNR value for sampled waveforms. Must be
+                specified if `waveform_dataset` is not `None`.
+            glitch_sampler:
+                Object for sampling glitches for insertion (not injection)
+                into background samples at data loading time. Must
+                include a `.sample` method. If passed a string, should be
+                a path to an HDF5 file containing glitches from each
+                interferometer contained in `"H1_glitches"` and
+                `"L1_glitches"` dataset keys. Data should be
+                sampled at rate `sample_rate`. Glitch triggers
+                should sit in the middle of the time axis of
+                the array containing the glitch timeseries.
+                Glitches will be randomly sampled and inserted in
+                place of the corresponding interferometer channel
+                background at data-loading time. If left as `None`,
+                no glitches will be inserted at data loading time
+                and `glitch_frac` should be 0.
+            glitch_frac:
+                The fraction of each batch that should consist
+                of inserted glitches, marked as `0.` in the
+                target tensor produced during iteration
+            device:
+                The device on which to host all the relevant
+                torch tensors.
+        """
+
+        # sanity check our fractions
+        assert 0 <= waveform_frac <= 1
+        assert 0 <= glitch_frac <= 1
+
+        self.sample_rate = sample_rate
+        self.kernel_size = int(kernel_length * sample_rate)
+        self.batch_size = batch_size
+        self.batches_per_epoch = batches_per_epoch
+        self.device = device
+
+        # load in the background data and build time-domain
+        # filters for whitening using them
+        self.hanford_background, hanford_asd, t0 = _load_background(
+            hanford_background, sample_rate, device
+        )
+        self.livingston_background, livingston_asd, t0 = _load_background(
+            livingston_background, sample_rate, device
+        )

```

Comment:
I assume these background files will contain real data. How many seconds of data are we using? Are we using all of this data to calculate the asd? Are we using the same asd to whiten all samples? 

Not super relevant to code, but important questions I've had that I figured should document somewhere.

Reply:
So I think ideally this implementation should be agnostic to those details, subject to the constraint that it's sufficiently small data to fit into (GPU) memory at once, so O(1GB). Which data and how much of it will change over time, but starting somewhere on the order of ~16K seconds might be ok to get started? Idk this is probably something good to discuss as part of #27 

But yeah the data will be real and all of it will be used to calculate the ASD, which will be used to whiten all samples both during training _and_ at inference time (the nice thing about structuring the whitening in Torch code is that we can save this step as part of the model so that it gets performed on the server at inference time).

Reply:
Got it. Just wanted to bring this up because the ASD is quite variable. Whitening an entire training set w/ the same asd may turn out to be problematic, although as a first pass this should be totally fine.

Reply:
Yeah I think the idea in the context of DL is just to use this as a normalization step, and in that case you just want to make sure you're applying the same normalization to test-time data as you did during training so that the statistics stay the same. If the statistics of your input data have shifted so dramatically that this whitening is no longer helpful, that's a problem but it isn't necessarily specific to whitening: you just have to retrain your network because the distribution you optimized for is no longer the one you're sampling from.

We could do an ifo-wise 0-mean unit-variance standardization instead, but this might be more natural in this context? Idk definitely something to keep an eye on